Improve service documentation (#129)
eclipse-velocitas/velocitas-docs: content/en/docs/tutorials/grpc_service_generation/create_server.md

---
title: "Create a server"
date: 2024-06-07T14:02:00+05:30
weight: 20
description: >
  Learn how to create a server for a service definition.
---

# Introduction

This example is based on that you have used the [Velocitas App C++ Template](https://github.com/eclipse-velocitas/vehicle-app-cpp-template) to create a new repository,

> and now you want to modify it to act as a client.

Reviewer: see above + it should act as a server not client

The example files can also be found in the [Github repository](https://github.com/eclipse-velocitas/velocitas-docs/examples/grpc_server).

## Velocitas components

Dependning on how you intend to deploy the Application and Databroker the number of Velocitas components required varies.
Below is the minimum set needed in `.velocitas.json` if deploying Databroker on localhost.

```json
    "components": [
        "devcontainer-setup",
        "vehicle-signal-interface",
        "grpc-interface-support",
        "sdk-installer",
        "build-system"
    ],
```

## App configuration

> In the `AppManifest.json` file you need to specify which interfaces your App want to use or provide.

Reviewer: ```suggestion
In the `AppManifest.json` file you need to specify which interfaces your server will provide.
```

In this case it declares that it want to implement everything from the Seats service defined in `seats.proto`,

> and that it want write access to the VSS signal `Vehicle.Cabin.Seat.Row1.DriverSide.Position`.

Reviewer: ```suggestion
In this case it declares to implement everything from the Seats service defined in `seats.proto` and that it needs write access to the VSS signal `Vehicle.Cabin.Seat.Row1.DriverSide.Position`.
```

```json

    "manifestVersion": "v3",
    "name": "SampleApp",
    "interfaces": [
        {
            "type": "grpc-interface",
            "config": {
                "src": "https://raw.githubusercontent.com/eclipse-kuksa/kuksa-incubation/0.4.0/seat_service/proto/sdv/edge/comfort/seats/v1/seats.proto",

>                 "provided": { }

Reviewer: don't you need to fill the array with the actual method names?

Author: Does not seem to be needed.

From the documentation at https://eclipse.dev/velocitas/docs/concepts/development_model/vehicle_app_manifest/interfaces/grpc_interface/#configuration-structure it actually seems that you cannot even give more detailed information.

Here I do not know how GRPC works. By default today we generate all methods returning `UNIMPLEMENTED`.
Does gRPC (C++ code) accepts that some bodies are missing? I assume we could adapt the generation so that if you list only method A out of A-C, then empty methods for B-C that return `UNIMPLEMENTED` would be hidden somewhere inside the SDK, but I do not know if that really would be beneficial.

Reviewer: yeah should be possible to only generate the stubs which are actually needed, but if it is working like this it is fine for now. let us only document what is currently there and not add additional features :)

            }
        },
        {
            "type": "vehicle-signal-interface",
            "config": {
                "src": "https://github.com/COVESA/vehicle_signal_specification/releases/download/v4.0/vss_rel_4.0.json",
                "datapoints": {
                    "required": [
                        {
                            "path": "Vehicle.Cabin.Seat.Row1.DriverSide.Position",
                            "access": "write"
                        }
                    ]
                }
            }
        }
    ]
}
```

## File Generation

When rebuilding the devcontainer with the configuration above two files will be automatically created:

* `SeatsServiceImpl.cpp`
* `SeatsServiceImpl.h`

> They contain the stub for the server. You need to fill in the actual implementation of the services.

Reviewer: ```suggestion
They contain the stubs for the server. You still need to fill in the actual implementation of the services.
```

## SeatsServiceImpl.cpp

In the file `SeatsServiceImpl.cpp` we need to implement the services.
In this example we only implement two of them.

> The example also show basic error handling.

Reviewer: ```suggestion
The example also shows basic error handling.
```

The business logic of the server is to wait for `Move` and `CurrentPosition` requests, and when received

> just forward the request to the Databroker and return the result.

Reviewer: ```suggestion
The business logic of the server is to wait for `Move` and `CurrentPosition` requests and when received just forward the request to the Databroker and return the result.
```

You do not need to add `SeatsServiceImpl.cpp` to `src/CMakeLists.txt`

> as the header includes the *.cpp file and th header file will be included from `Launcher.cpp`.

Reviewer: ```suggestion
You do not need to add `SeatsServiceImpl.cpp` to `src/CMakeLists.txt` as the header includes the *.cpp file and the header file will be included in the `Launcher.cpp`.
```

``` cpp
#include "SeatsServiceImpl.h"
#include <sdk/VehicleApp.h>
#include <sdk/VehicleModelContext.h>
#include <sdk/vdb/IVehicleDataBrokerClient.h>

namespace velocitas {

::grpc::Status SeatsService::Move(::grpc::ServerContext*                              context,
                                  const ::sdv::edge::comfort::seats::v1::MoveRequest* request,
                                  ::sdv::edge::comfort::seats::v1::MoveReply*         response) {
    (void)context;
    (void)request;
    (void)response;
    auto seat     = request->seat();
    auto location = seat.location();
    auto row      = location.row();
    auto pos      = location.index();

    std::cout << "Got Move Request!" << std::endl;

    try {
        auto status =

>             m_vehicleModel->Cabin.Seat.Row1.DriverSide.Position.set(seat.position().base())

Reviewer: why do you use that m_vehicleModel and not the standard Vehicle?

Author: It depends on what we mean with `standard`. This is the approach we use in:

https://github.com/eclipse-velocitas/vehicle-app-cpp-sdk/blob/main/examples/seat-adjuster/src/SeatAdjusterApp.h#L75

I did some testing forward and backward, I think I can use the other approach, will do a sanity check

Reviewer: I was more referring to our templates like here: https://github.com/eclipse-velocitas/vehicle-app-python-template/blob/5f2e678a86e909b2e9e583a2205c985a8b9fdf6b/app/src/main.py#L66

But also see now that we don't inherit from the sdks base vehicleapp...

---
title: "Create a server"
date: 2024-06-07T14:02:00+05:30
weight: 20
description: >
  Learn how to create a server for a service definition.
---

# Introduction

> This example assumes that you have used the [Velocitas App C++ Template](https://github.com/eclipse-velocitas/vehicle-app-cpp-template) to create a new repository and now want to modify it to be a grpc service client.

Reviewer: ```suggestion
This example assumes that you have used the [Velocitas App C++ Template](https://github.com/eclipse-velocitas/vehicle-app-cpp-template) to create a new repository and now want to modify it to be a grpc service server.
```

The example files can also be found in the [Github repository](https://github.com/eclipse-velocitas/velocitas-docs/examples/grpc_server).

## Velocitas components

Dependning on how you intend to deploy the Application and Databroker the number of Velocitas components required varies.
Below is the minimum set needed in `.velocitas.json` if deploying Databroker on localhost.

```json
    "components": [
        "devcontainer-setup",
        "vehicle-signal-interface",
        "grpc-interface-support",
        "sdk-installer",
        "build-system"
    ],
```

## App configuration

> In the `AppManifest.json` file you need to specify which interfaces your server will provide.

Reviewer: This line is not true, since we only have to specify that we are a provider in general.

Author: Really?

In the JSON file you need specify a list of explicit proto interfaces, right? It is not sufficient to say that we are a generic proto provider.

Reviewer: since the provide array is empty it does not directly say which interfaces it provides, we just say please generate the provider in general

Author: This is getting a bit philosophical. From a gRPC/Proto-pespective - what is actually the interface? Is it the service definition, or is each rpc/method in the service definition considered as an interface? Some quick googling did not give a clear answer :-)

To avoid this ambiguity we could rephrase to something like:

_In the `AppManifest.json` file you need to specify all gRPC services your application will use, either as a client or a server. For a server you need to specify `"provided": { }`. It is not needed and not supported to specify individual methods, as the generated code will include all methods defined in the gRPC service._

That way we avoid the ambiguous word _interface_, and instead state what I think we agree on - that you must list all gRPC services, but you do need (and cannot) to list gRPC methods/rpcs.

> In the `AppManifest.json` file you need to specify which interfaces your server will provide.

Reviewer: ```suggestion
In the `AppManifest.json` you need to specify that your server will provide the interfaces defined in the proto file.
```

> If it require access to signals in Databroker it must also be specified.

Reviewer: ```suggestion
If it requires access to signals from the Databroker, they must also be specified.
```

> In this case it declares to implement everything from the Seats service defined in `seats.proto` and that it needs write access to the VSS signal `Vehicle.Cabin.Seat.Row1.DriverSide.Position`.

Reviewer: ```suggestion
In this example the server declares to provide the interfaces from the Seats service defined in `seats.proto` and that it needs write access to the VSS signal `Vehicle.Cabin.Seat.Row1.DriverSide.Position`.
```

```json

    "manifestVersion": "v3",
    "name": "SampleApp",
    "interfaces": [
        {
            "type": "grpc-interface",
            "config": {
                "src": "https://raw.githubusercontent.com/eclipse-kuksa/kuksa-incubation/0.4.0/seat_service/proto/sdv/edge/comfort/seats/v1/seats.proto",
                "provided": { }
            }
        },
        {
            "type": "vehicle-signal-interface",
            "config": {
                "src": "https://github.com/COVESA/vehicle_signal_specification/releases/download/v4.0/vss_rel_4.0.json",
                "datapoints": {
                    "required": [
                        {
                            "path": "Vehicle.Cabin.Seat.Row1.DriverSide.Position",
                            "access": "write"
                        }
                    ]
                }
            }
        }
    ]
}
```

## File Generation

When rebuilding the devcontainer with the configuration above two files will be automatically created:

* `SeatsServiceImpl.cpp`
* `SeatsServiceImpl.h`

They contain the stubs for the server. You still need to fill in the actual implementation of the services.

## SeatsServiceImpl.h

> The service will access the vehicle model, in this tutorial we manage that by adding a private class variable `vehicle`
> to the file `SeatsServiceImpl.h`

Reviewer: ```suggestion
The service will access the vehicle model, in this tutorial we manage that by adding a private class variable `vehicle` to the file `SeatsServiceImpl.h`
```

You also need to adapt the constructor to be able to pass the vehicle object, otherwise it is currently not initialized in the example as far as I can see.

Author: It is a class variable (not a pointer), so initialized automatically when class  initialized, same construct like in https://github.com/eclipse-velocitas/vehicle-app-cpp-template/blob/main/app/src/SampleApp.h and https://github.com/eclipse-velocitas/vehicle-app-cpp-template/blob/main/app/src/SampleApp.cpp, or?

Reviewer: hmm yes you are right, I was thinking of another example we developed and there we had used and started several clients and service in one method. there it was better to pass the reference to the model instead of always initializing a new model.
But for this example it should be fine

In this example we only implement two of them.
The example also shows basic error handling.

The business logic of the server is to wait for `Move` and `CurrentPosition` requests and when received just forward the request to the Databroker and return the result.
You do not need to add `SeatsServiceImpl.cpp` to `src/CMakeLists.txt` as the header includes the *.cpp file and the header file will be included in the `Launcher.cpp`.

```cpp
#include "SeatsServiceImpl.h"
#include <sdk/VehicleApp.h>
#include <sdk/VehicleModelContext.h>
#include <sdk/vdb/IVehicleDataBrokerClient.h>

namespace velocitas {

::grpc::Status SeatsService::Move(::grpc::ServerContext*                              context,
                                  const ::sdv::edge::comfort::seats::v1::MoveRequest* request,
                                  ::sdv::edge::comfort::seats::v1::MoveReply*         response) {
    (void)context;
    (void)request;
    (void)response;
    auto seat     = request->seat();
    auto location = seat.location();
    auto row      = location.row();
    auto pos      = location.index();

    std::cout << "Got Move Request!" << std::endl;

    try {
        auto status =

>             vehicle.Cabin.Seat.Row1.DriverSide.Position.set(seat.position().base())->await();

Reviewer: don't you need to set it the position to a specific value?

Author: The client sets it to 75, the server sets it to the value it receives from the client.

I have not really tried to change the business logic of the "old" example, it should have the same features as before, even if they do not really makes sense always.

Reviewer: ah Github collapsed the lines where the seat was constructed from the request, sorry. should be fine

        std::cout << "Success!!" << std::endl;
        seat_position->set_base(seatPos);
        return ::grpc::Status(::grpc::StatusCode::OK, "");
    } catch (AsyncException& e) {
        std::cout << "Async Exception!" << std::endl;
        // This could typically be that Databroker is not running or not reachable
        // or that your Velocitas environment uses an API incompatible with what your Databroker
        // instance supports
        return ::grpc::Status(::grpc::StatusCode::OK, "");
    } catch (InvalidValueException& e) {
        std::cout << "Invalid Value!" << std::endl;
        // This could be given if Databroker has no current value for Position
        // (Have you set it manually using kuksa-client or databroker-cli?)
        return ::grpc::Status(::grpc::StatusCode::OK, "");
    }
}

} // namespace velocitas

```

## Launcher.cpp

In addition to the `SeatsServiceImpl.cpp` file we also need a file that starts the server, 
in this example called `Launcher.cpp`. It has very simple logic, it only starts the service and
waits until the service shuts down, if ever.

> You need to add `Launcher.cpp` to `src/CMakeLists.txt`.

Reviewer: as above in our standard template it is already there

``` cpp
#include "SeatsServiceImpl.h"
#include <sdk/middleware/Middleware.h>
#include <services/seats/SeatsServiceServerFactory.h>

#include "vehicle/Vehicle.hpp"

#include <memory>

using namespace velocitas;

int main(int argc, char** argv) {
    auto seatsImpl = std::make_shared<SeatsService>();

    velocitas::VehicleModelContext::getInstance().setVdbc(
        velocitas::IVehicleDataBrokerClient::createInstance("vehicledatabroker"));

    vehicle::Vehicle vehicle;

    auto seatServer = SeatsServiceServerFactory::create(Middleware::getInstance(), seatsImpl);

    seatServer->Wait();

    std::cout << "Waiting!" << std::endl;
    return 0;
}
```

## Building and Running

To (re-)build the App after changing the code you can use the [build script](https://github.com/eclipse-velocitas/vehicle-app-cpp-template/blob/main/build.sh).
As preparation for running you must also set two environment variables to indicate where the address/port of the

> server and the Databroker.

Reviewer: see above

        velocitas::IVehicleDataBrokerClient::createInstance("vehicledatabroker"));

    auto seatServer = SeatsServiceServerFactory::create(Middleware::getInstance(), seatsImpl);

    seatServer->Wait();

    std::cout << "Waiting!" << std::endl;
    return 0;
}
```

## Building and Running

To (re-)build the App after changing the code you can use the [build script](https://github.com/eclipse-velocitas/vehicle-app-cpp-template/blob/main/build.sh).
As preparation for running you must also set two environment variables to define where the address/port of the server and the Databroker.
The environment variable needs to be set in the same terminal as used for starting the application.

```bash
./build.sh
export SDV_SEATS_ADDRESS=grpc://127.0.0.1:5555
export SDV_VEHICLEDATABROKER_ADDRESS=grpc://127.0.0.1:55555
```

If Databroker is running and has a current value for the wanted signal, then everthing shall work when the client is started.
Output similar to below is expected when client has run once.

```bash
vscode ➜ /workspaces/erik_server_241021 (main) $ build/bin/app
2024-10-22 09:51:41, INFO  : Connecting to data broker service 'vehicledatabroker' via '127.0.0.1:55555'

> 2024-10-22 09:51:41, INFO  : Server sdv::edge::comfort::seats::v1::Seats listening on 127.0.0.1:5556

Reviewer: where is the 5556 port coming from? shouldn't it be 5555 as per env definition?

Author: Copy/paste error, port 5555 was not properly closed in one run, so I had to use another port, will address it.

    } catch (InvalidValueException& e) {
        std::cout << "Invalid Value!" << std::endl;
        // This could be given if Databroker has no current value for Position
        // (Have you set it manually using kuksa-client or databroker-cli?)
        return ::grpc::Status(::grpc::StatusCode::OK, "");
    }
}

} // namespace velocitas

```

## Launcher.cpp

You need to have a file that that starts the server,
in this example we modify the file `Launcher.cpp` that already exist in the [template](https://github.com/eclipse-velocitas/vehicle-app-cpp-template).
It has very simple logic, it only starts the service and
waits until the service shuts down, if ever.

``` cpp
#include "SeatsServiceImpl.h"
#include <sdk/middleware/Middleware.h>
#include <services/seats/SeatsServiceServerFactory.h>

#include "vehicle/Vehicle.hpp"

#include <memory>

using namespace velocitas;

int main(int argc, char** argv) {
    auto seatsImpl = std::make_shared<SeatsService>();

    velocitas::VehicleModelContext::getInstance().setVdbc(
        velocitas::IVehicleDataBrokerClient::createInstance("vehicledatabroker"));

    vehicle::Vehicle vehicle;

>     auto seatServer = SeatsServiceServerFactory::create(Middleware::getInstance(), seatsImpl);

Reviewer: here you need to pass the vehicle to the server

Author: We can actually remove the `vehicle` variable a few lines above as we have a private class variable (not pointer) in the server class, just like in https://github.com/eclipse-velocitas/vehicle-app-cpp-template/blob/main/app/src/SampleApp.h

Reviewer: see my other comment, for this example it is enough to initialize a new model

    seatServer->Wait();

    std::cout << "Waiting!" << std::endl;
    return 0;
}
```

## Building and Running

To (re-)build the App after changing the code you can use the [build script](https://github.com/eclipse-velocitas/vehicle-app-cpp-template/blob/main/build.sh).

> As preparation for running you must also set two environment variables to define where the address/port of the
> server and the Databroker.

Reviewer: ```suggestion
As preparation for running you must also set two environment variables to define where the address/port of the server and the Databroker.
```

please also add a note to start the server in the same terminal
